dsc_lib/src/functions/substring.rs

```diff
@@ -0,0 +1,233 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use super::Function;
+use crate::configure::context::Context;
+use crate::functions::{FunctionArgKind, FunctionCategory, FunctionMetadata};
+use crate::DscError;
+use rust_i18n::t;
+use serde_json::Value;
+use tracing::debug;
+
+#[derive(Debug, Default)]
+pub struct Substring {}
+
+impl Function for Substring {
+    fn get_metadata(&self) -> FunctionMetadata {
+        FunctionMetadata {
+            name: "substring".to_string(),
+            description: t!("functions.substring.description").to_string(),
+            category: vec![FunctionCategory::String],
+            min_args: 2,
+            max_args: 3,
+            accepted_arg_ordered_types: vec![
+                vec![FunctionArgKind::String],
+                vec![FunctionArgKind::Number],
+                vec![FunctionArgKind::Number],
+            ],
+            remaining_arg_accepted_types: None,
+            return_types: vec![FunctionArgKind::String],
+        }
+    }
+
+    fn invoke(&self, args: &[Value], _context: &Context) -> Result<Value, DscError> {
+        debug!("{}", t!("functions.substring.invoked"));
+
+        let string_to_parse = args[0].as_str().unwrap();
+        let start_index_value = args[1].as_i64().unwrap();
+
+        if start_index_value < 0 {
+            return Err(DscError::FunctionArg(
+                "substring".to_string(),
+                t!("functions.substring.startIndexNegative").to_string(),
+            ));
+        }
+
+        let start_index = usize::try_from(start_index_value).map_err(|_| {
+            DscError::FunctionArg(
+                "substring".to_string(),
+                t!("functions.substring.startIndexValueTooLarge").to_string(),
+            )
+        })?;
+        let string_length = string_to_parse.chars().count();
```

Comment:
Shouldn't `len()` just work here?

```suggestion
        let string_length = string_to_parse.len();
```

Reply:
It probably does, but if `len()` was used, Unicode characters are not preserved. 

```rust
let s = "héllo";
println!("len(): {}", s.len());           // 6 bytes (é takes 2 bytes)
println!("chars().count(): {}", s.chars().count()); // 5 characters
```